### PR TITLE
Change Profile buffer into a ring buffer

### PR DESCRIPTION
**Pull request by jpsamaroo**
Users of Profile may be interested in running the profiler continuously and without stopping. In the case of Dagger, it's possible that multiple tasks may all want the profiler running during their execution, so we only stop the profiler if Dagger isn't running any tasks. It's thus possible that we want the profiler running for longer than it would take to fill up the profile buffer. In the old implementation, the buffer would fill up, profiling would stop, and new traces would be lost. Additionally, sampling from the profiler with `Profile.fetch` while the profiler is running would often produce partially-written frames, requiring such continuously-profiling users to clean up such frames manually.

The new implementation treats the profile buffer as a ring buffer instead. `Profile.fetch` will now start at the `len_data()` pointer offset, read until the end of the buffer, and then read the rest from the beginning to just before `len_data()`. Any zero frames at the beginning of the copied buffer are truncated, so we end up with the result we would expect without a ring buffer.

Additionally, `Profile.fetch` now takes the same locks as the profiler while copying out data from the buffer, so sampling while profiling is running should now work correctly, without broken frames.

Marking this as a draft since I still need to test this locally with Dagger.

### CODE REVIEW

File: stdlib/Profile/src/Profile.jl [hunk 1]
```diff
@@ -505,15 +503,26 @@ By default metadata such as threadid and taskid will be stripped. Set `include_m
...
+        idx = findfirst(!iszero, data)
+        if idx !== nothing
+            data = data[idx:end]
+        else
+            data = Vector{UInt}(undef, 0)
+        end
     end
-    data = Vector{UInt}(undef, len)
-    GC.@preserve data unsafe_copyto!(pointer(data), get_data_pointer(), len)
-    if include_meta || isempty(data)
```
Comment (jpsamaroo) [inline stdlib/Profile/src/Profile.jl:516]: I removed the `isempty(data)` check assuming that `data` could never be empty anymore, but I'm not sure if it's possible to have a 0-sized profile buffer?

File: stdlib/Profile/src/Profile.jl [hunk 2]
```diff
@@ -505,15 +503,26 @@ By default metadata such as threadid and taskid will be stripped. Set `include_m
...
-                 before your program finished. To profile for longer runs, call
-                 `Profile.init()` with a larger buffer and/or larger delay."""
+    GC.@preserve data begin
+        data_ptr = get_data_pointer()
+        tail = maxlen - len
+        unsafe_copyto!(pointer(data), data_ptr+sizeof(UInt)*len, tail)
+        unsafe_copyto!(pointer(data)+sizeof(UInt)*tail, data_ptr, len)
+    end
+    ccall(:jl_profile_unlock, Cvoid, ())
+    if data[1] == 0
```
Comment (jpsamaroo) [inline stdlib/Profile/src/Profile.jl:517]: This is wrong: we need to make sure that we're not truncating a valid block end at the beginning of the buffer.

File: src/signals-unix.c [hunk 3]
```diff
@@ -766,51 +766,45 @@ static void *signal_listener(void *arg)
...
+                // and restore from the SEGV handler if anything happens.
+                jl_jmp_buf *old_buf = jl_get_safe_restore();
+                jl_jmp_buf buf;
+
+                jl_set_safe_restore(&buf);
+                if (jl_setjmp(buf, 0)) {
+                    jl_safe_printf("WARNING: profiler attempt to access an invalid memory location\n");
+                } else {
+                    // Get backtrace data
+                    bt_size_cur += rec_backtrace_ctx((jl_bt_element_t*)bt_data_prof + bt_size_cur,
```
Comment (jpsamaroo) [inline src/signals-unix.c:779]: Needs to be fixed like macOS and Windows.

**Comment by jpsamaroo** [comment]
Something to consider: should I pass a stack-allocated buffer to `rec_backtrace_ctx` to ensure we capture enough frames when at the end of the ring buffer? If so, how large should that buffer be?

**Comment by jpsamaroo** [comment]
I guess instead of writing out a frame that may wrap around to the beginning of the buffer, we should probably instead jump to the beginning of the buffer, to ensure we don't get partial frames when copying data out.

**Comment by IanButterworth** [comment]
I'm not sure that's necessary, why couldn't a block wrap around?

I guess a challenge is knowing whether part of a block has been overwritten and this incomplete?  Perhaps the easy way to handle it would be to always ignore the first block after the last write point, if it's wrapped?

The new `Profile.is_block_end` function should be useful for this.

**Comment by jpsamaroo** [comment]
Superseded by https://github.com/JuliaLang/julia/pull/53835

**Pull request closed without merging**